gto/git_utils.py

```diff
@@ -82,140 +92,72 @@ def _call_commit_push(
+                f"Make sure your local repository is in sync with the remote."
+            )
 
 
-def git_add_and_commit_all_changes(repo: Union[str, git.Repo], message: str) -> None:
-    repo = read_repo(repo)
-    tracked, untracked = _get_repo_changed_tracked_and_untracked_files(repo=repo)
-    if len(tracked) + len(untracked) > 0:
+def git_add_and_commit_all_changes(scm: Git, message: str) -> None:
+    staged, unstaged, untracked = scm.status()
+    if staged or unstaged or untracked:
+        files = list(chain(staged, unstaged, untracked))
         logging.debug("Adding to the index the untracked files %s", untracked)
-        logging.debug("Add and commit changes to files %s", tracked + untracked)
-        repo.index.add(items=tracked + untracked)
-        repo.index.commit(message=message)
+        logging.debug("Add and commit changes to files %s", files)
+        scm.add(files)
+        scm.commit(message)
 
 
-def read_repo(repo: Union[str, git.Repo], search_parent_directories=False) -> git.Repo:
-    if isinstance(repo, Repo):
-        return repo
-    try:
-        return git.Repo(repo, search_parent_directories=search_parent_directories)
-    except (InvalidGitRepositoryError, NoSuchPathError) as e:
-        raise NoRepo(repo) from e
+def _reset_repo_to_head(scm: Git) -> None:
+    if scm.stash.push(include_untracked=True):
+        scm.stash.drop()
 
 
-@contextmanager
-def stashed_changes(repo: Union[str, git.Repo], include_untracked: bool = False):
-    repo = read_repo(repo)
-    if len(repo.refs) == 0:
-        raise RuntimeError(
-            "Cannot stash because repository has no ref. Please create a first commit."
-        )
-
-    tracked, untracked = _get_repo_changed_tracked_and_untracked_files(repo=repo)
-
-    stash_arguments = ["push"]
-    if include_untracked:
-        stash_arguments += ["--include-untracked"]
-    else:
-        untracked = []
-
+def has_remote(scm: Git, remote: str = "origin") -> bool:
     try:
-        if len(tracked + untracked) > 0:
-            repo.git.stash(stash_arguments)
-        yield tracked, untracked
-    finally:
-        if len(tracked + untracked) > 0:
-            repo.git.stash("pop")
-
-
-def _reset_repo_to_head(repo: Union[str, git.Repo]) -> None:
-    repo = read_repo(repo)
-    repo.git.stash(["push", "--include-untracked"])
-    repo.git.stash(["drop"])
-
-
-def _get_repo_changed_tracked_and_untracked_files(
-    repo: Union[str, git.Repo],
-) -> Tuple[List[str], List[str]]:
-    repo = read_repo(repo)
-    return [item.a_path for item in repo.index.diff(None)], repo.untracked_files
-
-
-def _turn_args_into_kwargs(
-    f: Callable, args: tuple, kwargs: Dict[str, object]
-) -> Dict[str, object]:
-    kwargs_complement = {
-        k: args[i]
-        for i, k in enumerate(inspect.signature(f).parameters.keys())
-        if i < len(args)
-    }
-    kwargs.update(kwargs_complement)
-    return kwargs
+        scm.validate_git_remote(remote)
```

Comment:
does this actually work (for [all backends](https://github.com/iterative/scmrepo/blob/63a1b4b96fce326975614324a3f39087f4d259f2/src/scmrepo/git/backend/pygit2/__init__.py#L927C9-L927C28))?
The original `is_url_of_remote_repo` just validated the url which safe and lightweight. Is there a reason validate something here and is it safe?

Reply:
It's not implemented in pygit yet, but it is implemented for dulwich. `scmrepo` will automatically use the backend that has an implementation for any given `scm.func()` call.

The issue here was that the original `is_url_of_remote_repo` wasn't actually valid. The regex based string matching that was used before really only works for identifying github/gitlab URLs. It's possible to have Git remotes that did not match the regex.

Reply:
Got it, I don't think we ever hit a usecase outside of remote github/gitlab so if the generalization was not air-tight it might've been missed or not impactful